File: educational-core/src/com/jetbrains/edu/learning/actions/NextTaskAction.java
package com.jetbrains.edu.learning.actions;

import com.intellij.icons.AllIcons;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import com.jetbrains.edu.learning.navigation.StudyNavigator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class NextTaskAction extends TaskNavigationAction {

  public static final String ACTION_ID = "Educational.NextTask";
  public static final String SHORTCUT = "ctrl pressed PERIOD";

  public NextTaskAction() {
    super("Next Task", "Navigate to the next task", AllIcons.Actions.Forward);
  }

  @Override
  protected Task getTargetTask(@NotNull final Task sourceTask) {
    return StudyNavigator.nextTask(sourceTask);
  }

  @NotNull
  @Override
  public String getActionId() {
    return ACTION_ID;
  }

  @Nullable
  @Override
  public String[] getShortcuts() {
    return new String[]{SHORTCUT};
  }
}

File: educational-core/src/com/jetbrains/edu/learning/actions/NextTaskAction.kt
package com.jetbrains.edu.learning.actions

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.Project
import com.jetbrains.edu.learning.courseFormat.tasks.Task
import com.jetbrains.edu.learning.navigation.NavigationUtils
import com.jetbrains.edu.learning.stepik.hyperskill.openNextActivity
import com.jetbrains.edu.learning.taskDescription.ui.check.CheckPanel
import org.jetbrains.annotations.NonNls

class NextTaskAction : TaskNavigationAction() {

  override fun getTargetTask(sourceTask: Task): Task? = NavigationUtils.nextTask(sourceTask)

  override fun update(e: AnActionEvent) {
    if (CheckPanel.ACTION_PLACE == e.place) {
      //action is being added only in valid context
      //no project in event in this case, so just enable it
      return
    }
    super.update(e)
  }

  override fun getCustomAction(task: Task): ((Project, Task) -> Unit)? {
    return if (NavigationUtils.isLastHyperskillProblem(task)) ::openNextActivity else null
  }

  companion object {
    @NonNls
    const val ACTION_ID = "Educational.NextTask"
  }
}
